### DISCUSSION TITLE
how do you create these crates ?

### DISCUSSION THREAD

**Discussion by HosMercury** [maxcountryman/tower-sessions#115] [Q&A] [answered]
From where to learn to do things like these? i am curious

**Comment by maxcountryman** [comment]
Are you asking more generally about crates like `tower-sessions` or do you mean something specific like how do you create a session store crate?

**Reply by HosMercury** [reply]
I am asking generally how to be professional like you, doing some crates that community uses

**Reply by maxcountryman** [reply, marked as answer]
It's a broad question and I wish I had more actionable answer for you, but I'd attribute a large part of my approach to years spent working with open source projects. In other words, I don't know of a better way of learning than by doing and there's really no shortcut I've found. I've learned basically everything I know by observing how other projects worked, attempting to contribute where I thought it might be valuable, and ultimately applying what I'd learned to my own projects. I'd also point out that my example might not be the best to follow: I maintain one popular package for another language but myself have written libraries which have had only mild success (in terms of usage, if that's your metric). Others have been far more successful than I have been at identifying larger needs and then providing tools to tackle those challenges and they might have more insight than I do into their approach. Moreover, those same folks have done a far better job than I have with any of my projects in marketing them and building a community around them. Both incredibly important elements of success, in my view. Anyway, I hope this helps in some way--my best advice is to observe, try to learn from those observations, and then put in the time (many years) to develop a sense for how to approach open source work.

**Reply by HosMercury** [reply]
Thanks
